File: app/src/main/java/com/hijamoya/example/dagger/injection/ConductorInjection.java
package com.hijamoya.example.dagger.injection;

import android.app.Activity;

import com.bluelinelabs.conductor.Controller;

import dagger.android.DispatchingAndroidInjector;
import dagger.internal.Preconditions;

public class ConductorInjection {

    public static void inject(Controller controller) {
        Preconditions.checkNotNull(controller, "controller");
        HasControllerInjector hasDispatchingControllerInjector = findControllerInjector(controller);
        DispatchingAndroidInjector<Controller> controllerInjector = hasDispatchingControllerInjector
                .controllerInjector();
        Preconditions.checkNotNull(controllerInjector, "%s.controllerInjector() returned null",
                hasDispatchingControllerInjector.getClass().getCanonicalName());
        controllerInjector.inject(controller);
    }

    private static HasControllerInjector findControllerInjector(Controller controller) {
        Controller parentController = controller;
        do {
            if ((parentController = parentController.getParentController()) == null) {
                Activity activity = controller.getActivity();
                if (activity instanceof HasControllerInjector) {
                    return (HasControllerInjector) activity;
                }
                if (activity.getApplication() instanceof HasControllerInjector) {
                    return (HasControllerInjector) activity.getApplication();
                }
                throw new IllegalArgumentException("No injector was found");
            }
        } while (!(parentController instanceof HasControllerInjector));
        return (HasControllerInjector) parentController;
    }

}

File: app/src/main/kotlin/com/hijamoya/example/dagger/injection/ConductorInjection.kt
package com.hijamoya.example.dagger.injection

import com.bluelinelabs.conductor.Controller
import dagger.internal.Preconditions

class ConductorInjection {

    companion object {
        fun inject(controller: Controller) {
            Preconditions.checkNotNull(controller, "controller")
            val hasDispatchingControllerInjector = findControllerInjector(controller)
            val controllerInjector = hasDispatchingControllerInjector
                    .controllerInjector()
            Preconditions.checkNotNull(controllerInjector,
                    "%s.controllerInjector() returned null",
                    hasDispatchingControllerInjector.javaClass.canonicalName)
            controllerInjector.inject(controller)
        }

        private fun findControllerInjector(controller: Controller): HasControllerInjector {
            var parentController: Controller? = controller
            do {
                parentController = parentController?.parentController
                if (parentController == null) {
                    val activity = controller.activity
                    if (activity is HasControllerInjector) {
                        return activity
                    }
                    if (activity!!.application is HasControllerInjector) {
                        return activity.application as HasControllerInjector
                    }
                    throw IllegalArgumentException("No injector was found")
                }
            } while (parentController !is HasControllerInjector)
            return parentController
        }
    }
}
